torch/testing/_internal/common_optimizers.py

```diff
@@ -807,6 +811,51 @@ def optim_error_inputs_func_sparseadam(device, dtype):
     return error_inputs
 
 
+def _get_optim_inputs_including_global_cliquey_kwargs(
+    device, dtype, optim_info, skip=()
+) -> List[OptimizerInput]:
+    """
+    Return a list of all configs for a given optimizer as a list of OptimizerInputs,
+    including configs that have supported global cliquey kwargs (foreach, fused,
+    differentiable) based on optim_info.supported_impls.
+
+    The configs (optim_inputs) returned by optim_info.optim_inputs_func(...)
+    intentionally do NOT include global cliquey kwargs to give flexibility to tests.
+    For example, testing correctness between toggling foreach on and off is now
+    trivial. That said, we sometimes want to test for all possible configs on an
+    optimizer including all supported flags, so this helper returns all optim inputs.
+    """
+    optim_inputs = optim_info.optim_inputs_func(device=device)
+
+    # fused is currently not supported on CPU
+    supported_impls = tuple(
+        x
+        for x in optim_info.supported_impls
+        if x not in skip and not (str(device) == "cpu" and x == "fused")
+    )
+
+    all_optim_inputs = []
+    for optim_input in optim_inputs:
+        # Add the base config where all the flags are False
+        base_kwargs = deepcopy(optim_input.kwargs)
+        if len(supported_impls) != 0:
+            for flag in supported_impls:
+                base_kwargs[flag] = False
+            all_optim_inputs.append(
+                OptimizerInput(params=None, kwargs=base_kwargs, desc=optim_input.desc)
+            )
+        # Add a config for when each of the global cliquey kwargs is True
```

Comment:
nit: can you leave a comment here mentioning that for now we don't support differentiable with either foreach or fused thus we don't need a product.

Reply:
But that's the definition of global cliquey! Should I reiterate that here + redirect to bigger note?